File: solutions/ruby/connect/1/connect.rb
class Board
  def initialize(board_lines)
    # Limpiamos los espacios de formato y dividimos cada fila en caracteres individuales
    @board = board_lines.map { |line| line.strip.split(/\s+/) }
    @height = @board.size
    @width  = @board.first ? @board.first.size : 0
  end

  def winner
    return 'X' if player_wins?('X')
    return 'O' if player_wins?('O')
    ''
  end

  private

  def player_wins?(player)
    # Definimos los puntos de partida y las condiciones de victoria de cada jugador
    if player == 'X'
      # De Izquierda a Derecha: empieza en col 0, gana al llegar a col máxima
      start_positions = (0...@height).map { |r| [r, 0] }.select { |r, c| @board[r][c] == 'X' }
      win_condition   = ->(r, c) { c == @width - 1 }
    else
      # De Arriba a Abajo: empieza en fila 0, gana al llegar a fila máxima
      start_positions = (0...@width).map { |c| [0, c] }.select { |r, c| @board[r][c] == 'O' }
      win_condition   = ->(r, c) { r == @height - 1 }
    end

    return false if start_positions.empty?

    # Ejecutamos un algoritmo de búsqueda (BFS) utilizando una cola
    queue = start_positions.dup
    visited = start_positions.to_h { |pos| [pos, true] }

    while queue.any?
      curr_r, curr_c = queue.shift

      return true if win_condition.call(curr_r, curr_c)

      # Buscamos los 6 vecinos hexagonales válidos del mismo jugador
      hex_neighbors(curr_r, curr_c).each do |nr, nc|
        next if visited[[nr, nc]]
        
        if @board[nr][nc] == player
          visited[[nr, nc]] = true
          queue << [nr, nc]
        end
      end
    end

    false
  end

  # Devuelve las coordenadas de los 6 vecinos geométricos en una red hexagonal
  def hex_neighbors(r, c)
    relative_coordinates = [
      [0, -1], [0, 1],   # Izquierda, Derecha
      [-1, 0], [-1, 1],  # Diagonales superiores
      [1, -1], [1, 0]    # Diagonales inferiores
    ]

    relative_coordinates.map { |dr, dc| [r + dr, c + dc] }.select do |nr, nc|
      # Nos aseguramos de retornar solo vecinos que estén dentro de los límites del tablero
      nr.between?(0, @height - 1) && nc.between?(0, @width - 1)
    end
  end
end

File: solutions/java/connect/1/src/main/java/Connect.java
import java.util.LinkedList;
import java.util.Queue;

class Connect {
    private final char[][] grid;
    private final int rows;

    public Connect(String[] board) {
        if (board == null || board.length == 0) {
            this.rows = 0;
            this.grid = new char[0][0];
            return;
        }
        this.rows = board.length;
        this.grid = new char[rows][];
        // Eliminamos los espacios iniciales y finales para trabajar solo con los caracteres del tablero
        for (int i = 0; i < rows; i++) {
            this.grid[i] = board[i].replaceAll("\\s+", "").toCharArray();
        }
    }

    public Winner computeWinner() {
        if (rows == 0 || grid[0].length == 0) {
            return Winner.NONE;
        }

        // Verificamos si el jugador O ganó (conecta de arriba a abajo)
        if (checkWin('O')) {
            return Winner.PLAYER_O;
        }
        // Verificamos si el jugador X ganó (conecta de izquierda a derecha)
        if (checkWin('X')) {
            return Winner.PLAYER_X;
        }
        
        return Winner.NONE;
    }

    private boolean checkWin(char player) {
        int cols = grid[0].length;
        Queue<int[]> queue = new LinkedList<>();
        boolean[][] visited = new boolean[rows][cols];

        // Inicializamos la búsqueda desde el borde correspondiente a cada jugador
        if (player == 'O') {
            // El jugador O empieza desde la fila superior (fila 0)
            for (int c = 0; c < cols; c++) {
                if (c < grid[0].length && grid[0][c] == 'O') {
                    queue.add(new int[]{0, c});
                    visited[0][c] = true;
                }
            }
        } else {
            // El jugador X empieza desde la columna izquierda (columna 0)
            for (int r = 0; r < rows; r++) {
                if (grid[r].length > 0 && grid[r][0] == 'X') {
                    queue.add(new int[]{r, 0});
                    visited[r][0] = true;
                }
            }
        }

        // Las 6 direcciones posibles en un tablero hexagonal
        // Arriba, Arriba-Derecha, Izquierda, Derecha, Abajo-Izquierda, Abajo
        int[] dr = {-1, -1, 0, 0, 1, 1};
        int[] dc = {0, 1, -1, 1, -1, 0};

        // Búsqueda en Anchura (BFS) para encontrar un camino conectado
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int r = curr[0];
            int c = curr[1];

            // Condición de victoria
            if (player == 'O' && r == rows - 1) {
                return true; // O llegó al borde inferior
            }
            if (player == 'X' && c == grid[r].length - 1) {
                return true; // X llegó al borde derecho
            }

            // Exploramos los 6 vecinos hexagonales
            for (int i = 0; i < 6; i++) {
                int nr = r + dr[i];
                int nc = c + dc[i];

                if (nr >= 0 && nr < rows && nc >= 0 && nc < grid[nr].length && !visited[nr][nc] && grid[nr][nc] == player) {
                    visited[nr][nc] = true;
                    queue.add(new int[]{nr, nc});
                }
            }
        }
        return false;
    }
}